File: 10200/main.c
#include <stdio.h>

#define MAX 10002

int main(int argc, char* argv[])
{
   int i, j, n;
   int prime[MAX];
   int prime_len;
   char is_prime[MAX];
   float result;

   for (i = 0; i < MAX; i++)
      is_prime[i] = (i & 1);

   is_prime[2] = 1;

   for (i = 3; i * i < MAX; i += 2) {
      for (j = i; j * i < MAX; j += 2) {
         is_prime[j * i] = 0;
      }
   }

   prime[0] = 2;
   prime_len = 1;
   for (i = 3; i < MAX; i += 2) {
      if (is_prime[i]) {
         prime[prime_len++] = i;
      }
   }

   for (i = 0; i < MAX; i++) {
      n = i * i + i + 41;
      if (n < MAX) {
         is_prime[i] = is_prime[n];
      } else {
         for (j = 0; j < prime_len; j++) {
            if (n < prime[j] || n % prime[j] == 0)
               break;
         }
         is_prime[i] = (j == prime_len || n == prime[j]);
      }
   }

   prime[0] = 1;
   for (i = 1; i < MAX; i++) {
      prime[i] = prime[i - 1] + is_prime[i];
   }

   while (scanf("%d%d", &i, &j) == 2) {
      result = (((prime[j] - prime[i]) + is_prime[i]) * 100.0) / (((j - i) + 1.0));
      printf("%.2f\n", result + 0.000001);
   }

   return 0;
}

File: 10200/Main.java
import java.io.*;
import java.util.*;

public class Main {
   public static void main(String args[]) throws IOException {
      String line;
      StringBuilder answer = new StringBuilder();
      BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
      boolean[] isPrime = new boolean[10002];
      ArrayList<Integer> prime = new ArrayList<Integer>();

      isPrime[2] = true;
      for (int i = 3; i < isPrime.length; i += 2) {
         isPrime[i] = true;
      }

      for (int i = 3; i * i < isPrime.length; i += 2) {
         for (int j = i; j * i < isPrime.length; j += 2) {
            isPrime[j * i] = false;
         }
      }

      prime.add(2);
      for (int i = 3; i < isPrime.length; i += 2) {
         if (isPrime[i]) {
            prime.add(i);
         }
      }

      for (int i = 0; i < isPrime.length; i++) {
         int n = i * i + i + 41;
         if (n < isPrime.length) {
            isPrime[i] = isPrime[n];
         } else {
            isPrime[i] = true;
            for (int j = 0; j < prime.size(); j++) {
               int p = prime.get(j);
               if (n < p) break;
               if (n % p == 0) {
                  isPrime[i] = (n == p);
                  break;
               }
            }
         }
      }

      prime.clear();
      prime.add(1);
      for (int i = 1; i < isPrime.length; i++) {
         prime.add((prime.get(i - 1) + (isPrime[i] ? 1 : 0)));
      }

      while ((line = in.readLine()) != null) {
         StringTokenizer tokenizer = new StringTokenizer(line);
         int a = Integer.parseInt(tokenizer.nextToken());
         int b = Integer.parseInt(tokenizer.nextToken());
         double result = (((prime.get(b) - prime.get(a)) + (isPrime[a] ? 1 : 0)) * 100.0) / (b - a + 1);

         answer.append(String.format("%.2f\n", result));
      }
      System.out.print(answer);
   }
}
